### DISCUSSION TITLE
Philosophy of coding the RDKit in C++ vs. Python

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#5837] [Development]
Hi, just trying to understand the philosophy for which RDKit code is to be written in C++ vs. Python. The README states 
"Core data structures and algorithms in C++". I noticed at least some (apparent) algorithms in Python, for example [TorsionFingerprints.py](https://github.com/rdkit/rdkit/blob/e110a5c47939762c9106c3c93d01acd98e48081c/rdkit/Chem/TorsionFingerprints.py) and [QED.py](https://github.com/rdkit/rdkit/blob/f9b47d907fe5e91c45fa08e8658959ce39d9aa76/rdkit/Chem/QED.py).

I'm asking because Python is the language I'm most comfortable coding in, so I'm wondering if there's an opportunity for me to contribute code to the RDKit.

**Comment by thegodone** [comment]
Hi this is all matter of speed and usage. If you implement a simple code for yourself or a small community use python if not moves to cpp.Envoyé de mon iPhoneLe 5 déc. 2022 à 00:58, Jeremy Monat ***@***.***> a écrit :﻿
Hi, just trying to understand the philosophy for which RDKit code is to be written in C++ vs. Python. The README states
"Core data structures and algorithms in C++". I noticed at least some (apparent) algorithms in Python, for example TorsionFingerprints.py and QED.py.
I'm asking because Python is the language I'm most comfortable coding in, so I'm wondering if there's an opportunity for me to contribute code to the RDKit.

**Reply by greglandrum** [reply]
I'm going to disagree a bit with Guillaume: for me the primary argument for having code in C++ is that we can then make it available everywhere the RDKit itself is available: Python, Java, C#, postgresql, JavaScript, etc.
Python code, on the other hand, is only useable from Python.

Having said that, the majority of RDKit users are working in Python. And, as you pointed out, there are already things that are implemented purely in Python. 

So if you want to contribute and you are most comfortable in Python: please feel free to contribute in Python! We can often port things to C++ later if that ends up seeming like a good idea.

**Reply by bertiewooster** [reply]
Great, thanks!
